sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/RowBasedKeyValueBatch.java

```diff
@@ -0,0 +1,322 @@
+import org.apache.spark.sql.types.*;
+import org.apache.spark.unsafe.memory.MemoryBlock;
+import org.apache.spark.unsafe.Platform;
+
+
+/**
+ * RowBasedKeyValueBatch stores key value pairs in contiguous memory region.
+ *
+ * Each key or value is stored as a single UnsafeRow. The format for each record looks like this:
+ * [4 bytes total size = (klen + vlen + 4)] [4 bytes key size = klen]
+ * [UnsafeRow for key of length klen] [UnsafeRow for Value of length vlen]
+ * [8 bytes pointer to next]
+ * Thus, record length = 8 + klen + vlen + 8
+ *
+ * RowBasedKeyValueBatch will automatically acquire new pages (MemoryBlock) when the current page
+ * is used up.
+ *
+ * TODO: making each entry more compact, e.g., combine key and value into a single UnsafeRow
+ */
+public final class RowBasedKeyValueBatch extends MemoryConsumer{
+  private static final int DEFAULT_CAPACITY = 1 << 16;
+  private static final long DEFAULT_PAGE_SIZE = 64 * 1024 * 1024;
+
+  private final StructType keySchema;
+  private final StructType valueSchema;
+  private final int capacity;
+  private int numRows = 0;
+
+  // Staging row returned from getRow.
+  final UnsafeRow keyRow;
+  final UnsafeRow valueRow;
+
+  // ids for current key row and value row being retrieved
+  private int keyRowId = -1;
+
+  // full addresses for key rows and value rows
+  private long[] keyOffsets;
+
+  // if all data types in the schema are fixed length
+  private boolean allFixedLength;
+  private int klen;
+  private int vlen;
+  private int recordLength;
+
+  private MemoryBlock currentAndOnlyPage = null;
+  private Object currentAndOnlyBase = null;
+  private long recordStartOffset;
+  private long pageCursor = 0;
+
+  public static RowBasedKeyValueBatch allocate(StructType keySchema, StructType valueSchema,
+                                               TaskMemoryManager manager) {
+    return new RowBasedKeyValueBatch(keySchema, valueSchema, DEFAULT_CAPACITY, manager);
+  }
+
+  public static RowBasedKeyValueBatch allocate(StructType keySchema, StructType valueSchema,
+                                               TaskMemoryManager manager, int maxRows) {
+    return new RowBasedKeyValueBatch(keySchema, valueSchema, maxRows, manager);
+  }
+
+  public int numRows() { return numRows; }
+
+  public void close() {
+    if (currentAndOnlyPage != null) {
+      freePage(currentAndOnlyPage);
+      currentAndOnlyPage = null;
+    }
+  }
+
+  private boolean acquireNewPage(long required) {
+    try {
+      currentAndOnlyPage = allocatePage(required);
+    } catch (OutOfMemoryError e) {
+      return false;
+    }
+    currentAndOnlyBase = currentAndOnlyPage.getBaseObject();
+    Platform.putInt(currentAndOnlyBase, currentAndOnlyPage.getBaseOffset(), 0);
+    pageCursor = 4;
+    recordStartOffset = pageCursor + currentAndOnlyPage.getBaseOffset();
+
+    return true;
+  }
+
+  private long getKeyOffsetForFixedLengthRecords(int rowId) {
+    return recordStartOffset + rowId * recordLength + 8;
```

Comment:
Let's explicitly cast `recordLength` as long to prevent inadvertant overflows while multiplying 2 integers